cdi-core/src/main/java/com/linkedin/cdi/connection/S3Connection.java

```diff
@@ -82,6 +92,25 @@ public WorkUnitStatus execute(WorkUnitStatus status) {
       if (StringUtils.isBlank(s3SourceV2Keys.getTargetFilePattern())) {
         status.setBuffer(wrap(files));
       } else {
+        boolean shouldUpload = getExtractorKeys().getPayloads().size() != 0;
+        // upload to S3 if payload is empty, otherwise download from S3
+        if (shouldUpload) {
+          Path path = new Path(getExtractorKeys().getPayloadsBinaryPath());
+          FileSystem fileSystem = path.getFileSystem(new Configuration());
+          FSDataInputStream fsDataInputStream = fileSystem.open(path);
+          long fileSize = fileSystem.getFileStatus(path).getLen();
+          String fileName = path.getName();
+          PutObjectRequest putObjectRequest =
+              PutObjectRequest.builder().bucket(s3SourceV2Keys.getBucket()).key(fileName).build();
+          RequestBody requestBody = RequestBody.fromInputStream(fsDataInputStream, fileSize);
+          PutObjectResponse putObjectResponse = s3Client.putObject(putObjectRequest, requestBody);
+          ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(
+              JsonUtils.GSON.toJson(putObjectResponse).getBytes(StandardCharsets.UTF_8));
+          status.setBuffer(byteArrayInputStream);
+          byteArrayInputStream.close();
```

Comment:
Re: streams
Either move them to finally block or use try-with-resources statement

Reply:
Actually I made a mistake here. The stream is supposed to be closed by Gobblin runtime not here. Will update